pkg/kubecost/allocation.go

```diff
@@ -267,35 +313,111 @@ func (parc ProportionalAssetResourceCost) Key(insertByNode bool) string {
 
 type ProportionalAssetResourceCosts map[string]ProportionalAssetResourceCost
 
-func (parcs ProportionalAssetResourceCosts) Insert(parc ProportionalAssetResourceCost, insertByNode bool) {
+func (parcs ProportionalAssetResourceCosts) Clone() ProportionalAssetResourceCosts {
+	cloned := ProportionalAssetResourceCosts{}
+
+	for key, parc := range parcs {
+		cloned[key] = parc.Clone()
+	}
+	return cloned
+}
+
+func (parcs ProportionalAssetResourceCosts) Insert(parc ProportionalAssetResourceCost, insertByNode, isAccumulation bool) {
 	if !insertByNode {
 		parc.Node = ""
 		parc.ProviderID = ""
 	}
 	if curr, ok := parcs[parc.Key(insertByNode)]; ok {
-		parcs[parc.Key(insertByNode)] = ProportionalAssetResourceCost{
+
+		CPUPercentage := curr.CPUPercentage + parc.CPUPercentage
+		GPUPercentage := curr.GPUPercentage + parc.GPUPercentage
+		RAMPercentage := curr.RAMPercentage + parc.RAMPercentage
+
+		toInsert := ProportionalAssetResourceCost{
 			Node:                       curr.Node,
 			Cluster:                    curr.Cluster,
 			ProviderID:                 curr.ProviderID,
-			CPUPercentage:              curr.CPUPercentage + parc.CPUPercentage,
-			GPUPercentage:              curr.GPUPercentage + parc.GPUPercentage,
-			RAMPercentage:              curr.RAMPercentage + parc.RAMPercentage,
+			CPUPercentage:              CPUPercentage,
+			GPUPercentage:              GPUPercentage,
+			RAMPercentage:              RAMPercentage,
 			NodeResourceCostPercentage: curr.NodeResourceCostPercentage + parc.NodeResourceCostPercentage,
+			CPUComponents:              append(curr.CPUComponents, parc.CPUComponents...),
+			GPUComponents:              append(curr.GPUComponents, parc.GPUComponents...),
+			RAMComponents:              append(curr.RAMComponents, parc.RAMComponents...),
 		}
+
+		if isAccumulation {
```

Comment:
This pattern definitely makes thing spicier. That said, this can of course work. But it may be more straightforward to have an implementation that uses purely cumulative units so that accumulation is the same as aggregation. (As we discussed on Slack, that would mean recording cumulative cost per-resource, along with total cost per-asset-resource, and summing those. Then dividing to yield a percentage only at the very end.

That may or may not be too much trouble for the short term. But conditionals like `isAccumulation` within these kinds of functions have historically caused headaches.